x/proof/keeper/msg_server_submit_proof_test.go

```diff
@@ -0,0 +1,1336 @@
+
+				// Create a claim with a merkle root derived from a relay
+				// response containing the wrong session ID.
+				claimMsg := newTestClaimMsg(t,
+					validSessionHeader.GetSessionId(),
+					supplierAddr,
+					appAddr,
+					service,
+					wrongResponseSessionIdMerkleRootBz,
+				)
+				_, err = srv.CreateClaim(ctx, claimMsg)
+				require.NoError(t, err)
+
+				// Construct new proof message using the valid session header,
+				// *not* the one used in the session tree's relay response.
+				return newTestProofMsg(t,
+					supplierAddr,
+					validSessionHeader,
+					wrongResponseSessionIdSessionTree,
+					expectedClosestMerkleProofPath,
+				)
+			},
+			expectedErr: status.Error(
+				codes.FailedPrecondition,
+				types.ErrProofInvalidRelay.Wrapf(
+					"session headers session IDs mismatch; expected: %q, got: %q",
+					validSessionHeader.GetSessionId(),
+					wrongSessionIdHeader.GetSessionId(),
+				).Error(),
+			),
+		},
+		{
+			desc: "relay request signature must be valid",
+			newProofMsg: func(t *testing.T) *types.MsgSubmitProof {
+				// Set the relay request signature to an invalid byte slice.
+				invalidRequestSignatureRelay := newEmptyRelay(validSessionHeader, validSessionHeader)
+				invalidRequestSignatureRelay.Req.Meta.Signature = invalidSignatureBz
+
+				// Ensure a valid relay response signature.
+				signRelayResponse(t, "supplier", supplierAddr, keyRing, invalidRequestSignatureRelay)
+
+				invalidRequestSignatureRelayBz, err := invalidRequestSignatureRelay.Marshal()
+				require.NoError(t, err)
+
+				// Construct a session tree with 1 relay with a session header containing
+				// a session ID that doesn't match the expected session ID.
+				invalidRequestSignatureSessionTree := newEmptySessionTree(t, validSessionHeader)
+
+				// Add the relay to the session tree.
+				err = invalidRequestSignatureSessionTree.Update([]byte{1}, invalidRequestSignatureRelayBz, 1)
+				require.NoError(t, err)
+
+				// Get the Merkle root for the session tree in order to construct a claim.
+				invalidRequestSignatureMerkleRootBz, err := invalidRequestSignatureSessionTree.Flush()
+				require.NoError(t, err)
+
+				// Create a claim with a merkle root derived from a session tree
+				// with an invalid relay request signature.
+				claimMsg := newTestClaimMsg(t,
+					validSessionHeader.GetSessionId(),
+					supplierAddr,
+					appAddr,
+					service,
+					invalidRequestSignatureMerkleRootBz,
+				)
+				_, err = srv.CreateClaim(ctx, claimMsg)
+				require.NoError(t, err)
+
+				// Construct new proof message derived from a session tree
+				// with an invalid relay request signature.
+				return newTestProofMsg(t,
+					supplierAddr,
+					validSessionHeader,
+					invalidRequestSignatureSessionTree,
+					expectedClosestMerkleProofPath,
+				)
+			},
+			expectedErr: status.Error(
+				codes.FailedPrecondition,
+				types.ErrProofInvalidRelayRequest.Wrapf(
+					"error deserializing ring signature: %s",
```

Comment:
Do we have unit tests where the relay signature is formatted correctly but comes from the incorrect signer?

Reply:
Adding the following in the code:

```go
	{
			desc: "relay request signature is valid but signed by an incorrect application",
			newProofMsg: func(t *testing.T) *types.MsgSubmitProof {
				t.Skip("TODO_TECHDEBT(@bryanchriswhite): Implement this")
				return nil
			},
		},
```